autopropose.py

```diff
@@ -1,71 +1,138 @@
-import os
+import sys
 import time
 import logging
 import subprocess
+import contextlib
 
+from typing import (
+    Dict,
+    Tuple,
+    Iterator,
+)
 
+from ecdsa import SigningKey
+from ecdsa.curves import SECP256k1
+
+
+logging.basicConfig(level=logging.INFO, format='%(levelname)s %(message)s')
 LOGGER = logging.getLogger()
 
 
-def parse_conf(contents):
+def generate_key_pair_hex() -> Tuple[str, str]:
+    signing_key = SigningKey.generate(curve=SECP256k1)
+    verifying_key = signing_key.get_verifying_key()
+    signing_key_hex = signing_key.to_string().hex()
+    verifying_key_hex = "04" + verifying_key.to_string().hex()
+    return (signing_key_hex, verifying_key_hex)
+
+
+def generate_keypair() -> Tuple[str, str]:
+    (private_key, public_key) = generate_key_pair_hex()
+    return (private_key, public_key)
+
+
+def parse_conf(contents: str) -> Dict[str, str]:
     result = {}
     for line in contents.splitlines():
         if line == '':
             continue
         key, value = line.split('=', maxsplit=1)
-        value = value.strip('"')
-        value = value.strip("'")
+        value = value.strip('"').strip("'")
         result[key] = value
     return result
 
 
-def parse_conf_file(file_name):
+def parse_conf_file(file_name: str) -> Dict[str, str]:
     with open(file_name) as file:
         contents = file.read()
         return parse_conf(contents)
 
 
-def main():
-    LOGGER.setLevel(logging.INFO)
+def deploy(contract: str, private_key: str) -> None:
+    deploy_command = [
+        'docker', 'exec', 'rnode',
+        './bin/rnode',
+        '-c', '/var/lib/rnode-static/rnode.conf',
+        'deploy',
+        '--phlo-limit', '1000000000',
+        '--phlo-price', '1',
+        '--private-key', private_key,
+        contract,
+    ]
+
+    try:
+        LOGGER.info("Deploying %s", contract)
+        subprocess.check_output(deploy_command)
+    except subprocess.CalledProcessError as e:
+        LOGGER.error(e.cmd)
+        LOGGER.exception('deploy')
+        LOGGER.error(e.returncode)
+        LOGGER.error(e.stdout)
+        LOGGER.error(e.stderr)
+
+
+
+def propose() -> None:
+    propose_command = [
+        'docker',
+        'exec',
+        'rnode',
+        './bin/rnode',
+        '-c', '/var/lib/rnode-static/rnode.conf',
+        'propose',
+    ]
+
+    try:
+        LOGGER.info("Proposing...")
+        subprocess.check_output(propose_command)
+    except subprocess.CalledProcessError as e:
+        LOGGER.error(e.cmd)
+        LOGGER.exception('propose')
+        LOGGER.error(e.returncode)
+        LOGGER.error(e.stdout)
+        LOGGER.error(e.stderr)
+
+
+@contextlib.contextmanager
+def must_take_at_least(seconds: int) -> Iterator[None]:
+    started_at = int(time.monotonic())
+    try:
+        yield
+    finally:
+        finished_at = int(time.monotonic())
+        took = finished_at - started_at
+        if took < seconds:
+            wait_duration = seconds - took
+            time.sleep(wait_duration)
```

Comment:
What does this construction mean?
```
try:
    yield
finally:
    # ... actual code ...
```

Reply:
This how the context managers work.  You get an exception from `yield` it it gets raised within the body of `with` statement.